### PR TITLE
Get `CompactionJobInfo` from CompactFiles

### PR DESCRIPTION
**Pull request by DorianZheng**

### CODE REVIEW

File: tools/db_bench_tool.cc [hunk 1]
```diff
@@ -4578,7 +4578,7 @@ void VerifyDBFromDB(std::string& truth_db_name) {
       if (FLAGS_max_scan_distance != 0) {
         if (FLAGS_reverse_iterator) {
           GenerateKeyFromInt(
-              (uint64_t)std::max(0l, seek_pos - FLAGS_max_scan_distance),
+              (uint64_t)std::max(0ll, seek_pos - FLAGS_max_scan_distance),
```
Comment (ajkr) [inline tools/db_bench_tool.cc]: in travis this change is causing:

```
tools/db_bench_tool.cc:4581:73: error: no matching function for call to ‘max(long long int, int64_t)’
               (uint64_t)std::max(0ll, seek_pos - FLAGS_max_scan_distance),
```

It's very confusing to me. Maybe just use `static_cast<int64_t>(0)`

**Comment by riversand963** [comment]
@ajkr , wrt. reason why not trigger these two callbacks for `CompactFiles`. I found the following in code comment: https://github.com/facebook/rocksdb/blob/master/db/db_impl_compaction_flush.cc#L876:L888. Does this help?

File: tools/db_bench_tool.cc (hunk 1 above)
Comment (DorianZheng) [reply tools/db_bench_tool.cc]: sure

File: db/db_impl_compaction_flush.cc [hunk 2]
```diff
@@ -2672,6 +2642,43 @@ bool DBImpl::MCOverlap(ManualCompactionState* m, ManualCompactionState* m1) {
...
+      if (compaction_job_info->table_properties.count(fn) == 0) {
+        shared_ptr<const TableProperties> tp;
+        auto s = current->GetTableProperties(&tp, fmd, &fn);
+        if (s.ok()) {
+          compaction_job_info->table_properties[fn] = tp;
+        }
+      }
+    }
+  }
+  for (const auto newf : c->edit()->GetNewFiles()) {
```
[the remark this answers is not in the document]
Comment (DorianZheng) [reply db/db_impl_compaction_flush.cc]: yeap, it's better

File: java/src/main/java/org/rocksdb/WriteBatchWithIndex.java [hunk 3]
```diff
@@ -129,12 +129,10 @@ public WBWIRocksIterator newIterator() {
   public RocksIterator newIteratorWithBase(
       final ColumnFamilyHandle columnFamilyHandle,
       final RocksIterator baseIterator) {
-    RocksIterator iterator = new RocksIterator(
-        baseIterator.parent_,
-        iteratorWithBase(nativeHandle_,
-                columnFamilyHandle.nativeHandle_,
-                baseIterator.nativeHandle_));
-    //when the iterator is deleted it will also delete the baseIterator
+    RocksIterator iterator = new RocksIterator(baseIterator.parent_,
```
[the remark this answers is not in the document]
Comment (DorianZheng) [reply java/src/main/java/org/rocksdb/WriteBatchWithIndex.java]: I just perform `make format` and I've checked it, just some code format, so I think it's fine.

**Comment by DorianZheng** [comment]
@ajkr @riversand963 Sorry for the delay, just working on something else recently.

**Comment by DorianZheng** [comment]
@riversand963 PTAL

File: db/db_impl_compaction_flush.cc [hunk 4]
```diff
@@ -720,7 +722,8 @@ Status DBImpl::CompactFiles(const CompactionOptions& compact_options,
   assert(cfd);
 
   Status s;
-  JobContext job_context(0, true);
+  JobContext job_context(next_job_id_.fetch_add(1, std::memory_order_relaxed),
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/db_impl_compaction_flush.cc]: I'm under the impression that we use job_id 0 for application-created `job_context` and `next_job_id` for RocksDB-created one. Correct me if this is wrong @ajkr .

[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_impl_compaction_flush.cc]: Hm that's what I thought too but after checking it looks like we assign a real job number for manual flush. I'm ok with assigning a real job number here too. We can use "compaction_reason" to distinguish manual from automatic.

**Comment by riversand963** [comment]
LGTM again. Thanks @DorianZheng for the PR

**Pull request closed without merging**